File: tools/extra/fpgadiag/mux.py
#!/usr/bin/env python3
# Copyright(c) 2017, Intel Corporation
#
# Redistribution  and  use  in source  and  binary  forms,  with  or  without
# modification, are permitted provided that the following conditions are met:
#
# * Redistributions of  source code  must retain the  above copyright notice,
#  this list of conditions and the following disclaimer.
# * Redistributions in binary form must reproduce the above copyright notice,
#  this list of conditions and the following disclaimer in the documentation
#   and/or other materials provided with the distribution.
# * Neither the name  of Intel Corporation  nor the names of its contributors
#   may be used to  endorse or promote  products derived  from this  software
#   without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
# IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
# LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
# CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
# SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
# INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
# CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
from __future__ import absolute_import
from __future__ import print_function
import datetime
import itertools
import json
import math
import random
import tempfile
import subprocess
import os

max_threads = 2047       # 11 bits
max_count = 1048575      # 20 bits
max_stride = 4294967295  # 32 bit


class nlb_options(object):
    all_options = {
        "begin": list(range(1, 65535)),
        "end": list(range(2, 65535)),
        "cont": [True, False],
        "timeout-sec": list(range(1, 11)),
        "mode": ["read", "write"],
        "multi-cl": [1, 2, 4],
        "cache-policy": ["wrline-I", "wrline-M", "wrpush-I"],
        "cache-hint": ["rdline-I", "rdline-S"],
        "read-vc": ["vh0", "vh1", "vl0"],
        "write-vc": ["vh0", "vh1"],
        "wrfence-vc": ["vh0", "vh1"],
        "strided-access": list(range(1, 65)),
        "threads": list(range(1, 65)),
        "count": list(range(1, 100)),
        "stride": [int(math.pow(2, n)) for n in range(9)],
        "warm-fpga-cache": [True, False],
        "cool-fpga-cache": [True, False],
        "cool-cpu-cache": [True, False],
        "alt-wr-pattern": [True, False]
    }

    @property
    def keys(self):
        return ()

    def empty(self):
        while True:
            yield {}

    def random(self):
        while True:
            opts = {}
            opts.update(
                dict([(k, random.choice(self.all_options[k]))
                      for k in self.keys]))
            opts.update(self.const())
            if self.validate(opts):
                yield opts

    def iter(self):
        values = [self.all_options[k] for k in self.keys]
        for v in itertools.product(*values):
            opts = dict(list(zip(self.keys, v)))
            opts.update(self.const())
            if self.validate(opts):
                yield opts

    def const(self):
        return {'suppress-stats': True}

    def validate(self, opts):
        return True


class nlb0_options(nlb_options):

    @property
    def keys(self):
        return ("begin",
                "cont",
                "timeout-sec",
                "multi-cl",
                "cache-policy",
                "cache-hint",
                "read-vc",
                "write-vc",
                "wrfence-vc")

    def validate(self, opts):
        if int(opts["begin"]) % int(opts["multi-cl"]) > 0:
            return False
        return True


class nlb3_options(nlb_options):

    @property
    def keys(self):
        return ("begin",
                "end",
                "mode",
                "cont",
                "timeout-sec",
                "multi-cl",
                "strided-access",
                "cache-policy",
                "cache-hint",
                "read-vc",
                "write-vc",
                "wrfence-vc",
                "alt-wr-pattern",
                "warm-fpga-cache",
                "cool-fpga-cache",
                "cool-cpu-cache")

    def validate(self, opts):
        if opts["warm-fpga-cache"] and opts["cool-fpga-cache"]:
            return False
        if int(opts["begin"]) % int(opts["multi-cl"]) > 0:
            return False
        if int(opts["end"]) % int(opts["multi-cl"]) > 0:
            return False
        if opts.get("end") < opts.get("begin"):
            return False
        if opts.get("end", opts["begin"]) * opts.get("strided-access") > 65535:
            return False
        return True

    def const(self):
        return {'cont': False}


class mtnlb7_options(nlb_options):

    def const(self):
        return {"mode": "mt7", "suppress-stats": True}

    @property
    def keys(self):
        return ("threads",
                "count",
                "stride",
                "cache-policy",
                "cache-hint",
                "read-vc",
                "write-vc",
                )


class mtnlb8_options(nlb_options):

    def const(self):
        return {"mode": "mt8", "suppress-stats": True}

    @property
    def keys(self):
        return ("threads",
                "count",
                "stride",
                "cache-policy",
                "cache-hint",
                "read-vc",
                "write-vc",
                )

    def validate(self, options):
        if options["threads"] > max_threads:
            return False
        if options["count"] > max_count:
            return False
        if options["stride"] > max_stride:
            return False
        return True


def run_mux(muxfile, args, iteration=0):
    result = "PASS"
    stdout, stderr = None, None
    try:
        p = subprocess.Popen(
            '{} -c {} -s {}'.format(args.cmd, muxfile, args.socket_id),
            shell=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE)
        timedout = False
        stdout, stderr = p.communicate()
    except subprocess.CalledProcessError:
        result = "CRASH"
    else:
        if not timedout:
            result = "PASS" if p.returncode == 0 else "FAIL"
    if result == "PASS":
        stdout, stderr = None, None
    result = {"result": result,
              "filename": muxfile,
              "iteration": iteration,
              "stdout": stdout,
              "stderr": stderr}

    return result


def summarize(results):
    failures = 0
    passes = 0
    crashes = 0
    timeout = 0
    for r in results:
        if r["result"] == "PASS":
            passes += 1
        elif r["result"] == "FAIL":
            failures += 1
        elif r["result"] == "CRASH":
            crashes += 1
        elif r["result"] == "TIMEOUT":
            timeout += 1
    return {"overview": {"failures": failures,
                         "passes": passes,
                         "crashes": crashes,
                         "timeout": timeout,
                         "total": len(results)},
            "results": results}


def run(args):
    generator = args.generator
    n0 = nlb0_options()
    n3 = nlb3_options()
    mt7 = mtnlb7_options()
    mt8 = mtnlb8_options()
    engines = {"nlb0": n0,
               "nlb3": n3,
               "mtnlb7": mt7,
               "mtnlb8": mt8}
    results = []
    count = 1
    start_time = datetime.datetime.now()
    stop_on_fail = args.stop_on_fail
    while True:
        options = [{"name": n,
                    "app": n,
                    "config": next(getattr(engines[n], generator)())}
                   for n in args.engines]
        for opt in options:
            if opt["name"] in args.disable:
                opt["disabled"] = True
        if len(options) == 1:
            file_prefix = options[0]["name"]
            options = options[0]["config"]
        with tempfile.NamedTemporaryFile("w",
                                         prefix=file_prefix,
                                         suffix=".json",
                                         delete=False,
                                         dir="muxfiles") as fd:
            json.dump(options, fd, indent=4, sort_keys=True)
        print(count, "running with: ", fd.name)
        count += 1
        try:
            result = run_mux(fd.name, args, count)
        except KeyboardInterrupt:
            print("Stopping...")
            return results
            break
        else:
            results.append(result)

        if result["result"] != "PASS":
            print("tests failed with file: ", fd.name)
            if stop_on_fail:
                break
        if args.timeout is not None:
            total_sec = (datetime.datetime.now() - start_time).total_seconds()
            if total_sec > args.timeout:
                break
        if args.max_iterations is not None:
            if count >= args.max_iterations:
                break
    return results


def retry(args):
    results_file = args.retry
    results = []
    retries = []
    with open(results_file, "r") as fd:
        results = json.load(fd)
    for result in results["results"]:
        if result["result"] != "PASS":
            with open(result["filename"], "r") as fd:
                configs = json.load(fd)
            for config in configs:
                config["app"] = config["name"]
                if config["config"].get("mode") == "mt7":
                    config["name"] = "mtnlb7"
                elif config["config"].get("mode") == "mt8":
                    config["name"] = "mtnlb8"
                config["disable"] = True
            rerunfile = result["filename"].replace("random", "rerun")
            print(rerunfile)
            with open(rerunfile, "w") as fd:
                json.dump(configs, fd, indent=4, sort_keys=True)
            retries.append(run_mux(rerunfile, args, result["iteration"]))
    return retries


if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument("cmd")
    parser.add_argument(
        '-s', '--socket-id', default=0, choices=[0, 1], type=int)
    parser.add_argument(
        "-g", "--generator", default="iter", choices=["iter", "random"],
        help="generator type for generating options")
    parser.add_argument("-m", "--max-iterations", type=int,
                        help="maximum number of iterations to execute")
    parser.add_argument("-t", "--timeout", type=float,
                        help="stop running after timeout period has elapsed")
    parser.add_argument("--stop-on-fail", action="store_true", default=False,
                        help="stop running after first failure")
    parser.add_argument("--retry",
                        help="run failed cases from a results file")
    parser.add_argument("--engines", nargs="+",
                        default=["nlb0", "nlb3", "mtnlb7", "mtnlb8"])
    parser.add_argument("--disable", nargs="*", default=[])
    args = parser.parse_args()
    if not os.path.exists("muxfiles"):
        os.mkdir("muxfiles")

    if args.retry:
        retries = retry(args)
        with open(args.retry.replace("results", "retries"), "w") as fd:
            json.dump(summarize(retries), fd, indent=4, sort_keys=True)
    else:
        results = run(args)
        datestr = datetime.datetime.now().isoformat('_')
        fname = "mux-results-{}.json".format(datestr)
        with open(fname, "w") as fd_results:
            json.dump(summarize(results), fd_results, indent=4, sort_keys=True)

File: mybuild_docs/binaries/fpgadiag/stage/src/mux.cpp
// Copyright(c) 2017-2018, Intel Corporation
//
// Redistribution  and  use  in source  and  binary  forms,  with  or  without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of  source code  must retain the  above copyright notice,
//   this list of conditions and the following disclaimer.
// * Redistributions in binary form must reproduce the above copyright notice,
//   this list of conditions and the following disclaimer in the documentation
//   and/or other materials provided with the distribution.
// * Neither the name  of Intel Corporation  nor the names of its contributors
//   may be used to  endorse or promote  products derived  from this  software
//   without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

#include <memory>
#include <map>
#include <typeindex>
#include <chrono>
#include "nlb0.h"
#include "nlb3.h"
#include "nlb7.h"
#include "nlb_stats.h"
#include "fpga_app/accelerator_mux.h"
#include "log.h"
#include "utils.h"
#include "option.h"
#include "option_parser.h"
#include "nlb_stats.h"
#include "fpga_app/fpga_common.h"
#include <json-c/json.h>


using namespace intel::fpga;
using namespace intel::fpga::diag;
using namespace intel::fpga::nlb;
using namespace intel::utils;
using namespace opae::fpga::types;


std::map<std::string, accelerator_app::ptr_t(*)(const std::string &)> app_factory =
{
    { "nlb0",   [](const std::string & name){ UNUSED_PARAM(name); return accelerator_app::ptr_t(new nlb0());   }},
    { "nlb3",   [](const std::string & name){ UNUSED_PARAM(name); return accelerator_app::ptr_t(new nlb3());   }},
    { "nlb7",   [](const std::string & name){ UNUSED_PARAM(name); return accelerator_app::ptr_t(new nlb7());   }},
};

enum test_result
{
    incomplete = 0,
    pass,
    fail,
    error
};

static std::map<json_tokener_error, std::string> json_parse_errors =
{
    { json_tokener_success, "json_tokener_success" },
    { json_tokener_continue, "json_tokener_continue" },
    { json_tokener_error_depth, "json_tokener_error_depth" },
    { json_tokener_error_parse_eof, "json_tokener_error_parse_eof" },
    { json_tokener_error_parse_unexpected, "json_tokener_error_parse_unexpected" },
    { json_tokener_error_parse_null, "json_tokener_error_parse_null" },
    { json_tokener_error_parse_boolean, "json_tokener_error_parse_boolean" },
    { json_tokener_error_parse_number, "json_tokener_error_parse_number" },
    { json_tokener_error_parse_array, "json_tokener_error_parse_array" },
    { json_tokener_error_parse_object_key_name, "json_tokener_error_parse_object_key_name" },
    { json_tokener_error_parse_object_key_sep, "json_tokener_error_parse_object_key_sep" },
    { json_tokener_error_parse_object_value_sep, "json_tokener_error_parse_object_value_sep" },
    { json_tokener_error_parse_string, "json_tokener_error_parse_string" },
    { json_tokener_error_parse_comment, "json_tokener_error_parse_comment" },
    { json_tokener_error_size, "json_tokener_error_size" }
};
bool make_apps(const std::string & muxfile, std::vector<accelerator_app::ptr_t> & apps)
{
    std::ifstream inp;
    inp.open(muxfile.c_str());
    if (!inp.good())
    {
        return false;
    }
    inp.seekg(0, inp.end);
    size_t length = inp.tellg();
    inp.seekg(0, inp.beg);
    char jbuff[length];
    inp.read(jbuff, length);
    enum json_tokener_error err = json_tokener_success;
    struct json_object * root = json_tokener_parse_verbose(jbuff, &err);

    if (root == nullptr)
    {
        auto it = json_parse_errors.find(err);
        auto msg = it == json_parse_errors.end() ? "Unknown" : it->second;
        std::cerr << "ERROR parsing muxfile: " << msg << "\n";
        return false;
    }

    auto count = json_object_array_length(root);
    option_parser parser;
    for (int i = 0; i < count; ++i)
    {
        auto instance = json_object_array_get_idx(root, i);
        json_object *name_obj;
        json_object *config_obj;
        json_object *app_obj;
        if (json_object_object_get_ex(instance, "config", &config_obj) &&
            json_object_object_get_ex(instance, "name",   &name_obj) &&
            json_object_object_get_ex(instance, "app",    &app_obj))
        {
            std::string name = json_object_get_string(name_obj);
            std::string appname = json_object_get_string(app_obj);
            auto it = app_factory.find(appname);
            if (it != app_factory.end())
            {
                json_object *disabled_obj;
                auto app = it->second(name);
                if (json_object_object_get_ex(instance, "disabled", &disabled_obj))
                {
                    app->disabled(json_object_get_boolean(disabled_obj));
                }
                parser.parse_json(json_object_to_json_string(config_obj), app->get_options());
                apps.push_back(app);
            }
        }
        else
        {
            return false;
        }
    }
    return (apps.size() > 0);
}

int main(int argc, char* argv[])
{
    option_parser parser;
    option_map opts;
    uint64_t freq = 300000000;

    opts.add_option<bool>("help",             'h', option::no_argument,   "Show help", false);
    opts.add_option<uint8_t>("socket-id",     's', option::with_argument, "Socket id encoded in BBS");
    opts.add_option<uint8_t>("bus-number",    'B', option::with_argument, "Bus number of PCIe device");
    opts.add_option<uint8_t>("device",        'D', option::with_argument, "Device number of PCIe device");
    opts.add_option<uint8_t>("function",      'F', option::with_argument, "Function number of PCIe device");
    opts.add_option<std::string>("target",    't', option::with_argument, "Target platform. fpga or ase - default is fpga", "fpga");
    opts.add_option<std::string>("guid",      'G', option::with_argument, "GUID of accelerator to open");
    opts.add_option<std::string>("config",    'c', option::with_argument, "Path to JSON file containing mux sw apps");
    opts.add_option<bool>("suppress-header",  'H', option::no_argument, "Suppress header when showing results", false);
    opts.add_option<bool>("csv",              'V', option::no_argument, "Show results in CSV format", false);
    opts.add_option<uint64_t>("frequency",    'T', option::with_argument, "Clock frequency (used for bw measurements)", freq);
    logger log;
    log.set_level(logger::level::level_debug);
    if (!parser.parse_args(argc, argv, opts))
    {
        log.error("main") << "Error parsing command line arguments" << std::endl;
        opts.show_help("fpgamux", std::cout);
        return EXIT_FAILURE;
    };

    bool help = false;
    if (opts.get_value<bool>("help", help) && help){
        opts.show_help("fpgamux", std::cout);
        return EXIT_SUCCESS;
    }

    auto muxopt = opts.find("config");
    if (!muxopt || !muxopt->is_set() || !path_exists(muxopt->value<std::string>()))
    {
        log.error("main") << "Invalid or no muxfile specified" << std::endl;
        return EXIT_FAILURE;
    }
    opts.get_value<uint64_t>("frequency", freq);

    std::string muxfile = muxopt->value<std::string>();
    std::vector<accelerator_app::ptr_t> apps;

    if (!make_apps(muxfile, apps))
    {
        return EXIT_FAILURE;
    }

    auto guid_opt = opts.find("guid");
    if (guid_opt)
    {
        if (!guid_opt->is_set())
        {
            log.info("main") << "Setting guid from first app in mux configuration: " << apps[0]->afu_id() << "\n";
            *guid_opt = std::string(apps[0]->afu_id());
        }
        else
        {
            log.info("main") << "Using guid from options: " << guid_opt->value<std::string>() << "\n";
        }
    }

    std::vector<accelerator::ptr_t> acceleratorlist = accelerator::enumerate({ std::make_shared<option_map>(opts) });
    if (acceleratorlist.size() == 0)
    {
        log.error("accelerator::enumerate") << "Could not find accelerator with given filter\n";
        return EXIT_FAILURE;
    }

    std::string target = "fpga";
    opts.get_value("target", target);
    bool shared = target == "fpga";
    std::map<std::string, std::future<bool>> futures;
    std::map<std::string, test_result>       results;
    std::map<std::string, shared_buffer::ptr_t> dsm_list;
    size_t instance = 0;
    auto accelerator_ptr = acceleratorlist[0];
    if (!accelerator_ptr->open(shared))
    {
        log.error("accelerator::open") << "Error opening accelerator" << std::endl;
        return EXIT_FAILURE;
    }
    accelerator_ptr->reset();
    auto buffer = accelerator_ptr->allocate_buffer(GB(1));
    if (!buffer) {
        log.error("main") << "failed to allocate workspace and input/output buffers." << std::endl;
        return EXIT_FAILURE;
    }
    buffer_pool::ptr_t pool(new buffer_pool(buffer));
    auto dsm = pool->allocate_buffer(MB(2));
   // Read perf counters.
    fpga_cache_counters    start_cache_ctrs  = accelerator_ptr->cache_counters();
    fpga_fabric_counters   start_fabric_ctrs = accelerator_ptr->fabric_counters();
    for (auto app : apps)
    {
        accelerator::ptr_t muxed(new accelerator_mux(acceleratorlist[0], apps.size(), instance++, pool));
        app->assign(muxed->handle());
        if (!app->disabled() && app->setup())
        {
            results[app->name()] = test_result::incomplete;
            futures[app->name()] = app->run_async();
            dsm_list[app->name()] = app->dsm();
        }
    }

    std::map<std::string, test_result>::size_type complete = 0;
    dsm_tuple tpl;
    while(complete < results.size())
    {
        complete = 0;
        for(auto & kv : results)
        {
            if (kv.second == test_result::incomplete &&
                futures[kv.first].valid() &&
                futures[kv.first].wait_for(std::chrono::milliseconds(10)) == std::future_status::ready)
            {
                results[kv.first] = futures[kv.first].get() ? test_result::pass : test_result::fail;
                tpl += dsm_tuple(dsm_list[kv.first]);
                complete++;
            }
            else if(kv.second != test_result::incomplete)
            {
                complete++;
            }
        }
    }

    uint64_t cachelines = 0;
    for (auto app : apps)
    {
        if (!app->disabled())  cachelines += app->cachelines();
    }
    bool suppress_header = false;
    bool csv_format = false;
    opts.get_value<bool>("suppress-header", suppress_header);
    opts.get_value<bool>("csv-format", csv_format);
    // Read Perf Counters
    fpga_cache_counters  end_cache_ctrs  = accelerator_ptr->cache_counters();
    fpga_fabric_counters end_fabric_ctrs = accelerator_ptr->fabric_counters();
    tpl.put(dsm);
    std::cout << intel::fpga::nlb::nlb_stats(dsm,
                                             cachelines,
                                             end_cache_ctrs - start_cache_ctrs,
                                             end_fabric_ctrs - start_fabric_ctrs,
                                             freq,
                                             false,
                                             suppress_header,
                                             csv_format);

    for(auto & kv : results)
    {
        if (kv.second != test_result::pass)
        {
            return EXIT_FAILURE;
        }
    }
    return EXIT_SUCCESS;
}
